File: dmtp-py/dmtp/mtp/package.py
# -*- coding: utf-8 -*-
#
#   MTP: Message Transfer Protocol
#
#                                Written in 2020 by Moky <albert.moky@gmail.com>
#
# ==============================================================================
# MIT License
#
# Copyright (c) 2020 Albert Moky
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
# ==============================================================================

from .tlv import Data
from .protocol import DataType, TransactionID, Message, MessageFragment
from .header import Header


class Package(Data):
    """
        Optimal Length for UDP Package Body
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        MTU      : 576 bytes
        IP Head  : 20 bytes
        UDP Head : 8 bytes
        Header   : 12 bytes (excludes 'pages', 'offset' and 'bodyLen')
        Reserved : 24 bytes (includes 'pages', 'offset' and 'bodyLen')
    """
    OPTIMAL_BODY_LENGTH = 512

    def __init__(self, head: Header, body: Data = None, data=None):
        if data is None:
            if body is None:
                body = Data.ZERO
                data = head
            else:
                data = head.concat(body)
        elif body is None:
            body = Data.ZERO
        super().__init__(data=data)
        self.__head = head
        self.__body = body

    def __str__(self) -> str:
        clazz = self.__class__.__name__
        return '<%s: head=%s body_len=%d />' % (clazz, self.head, self.body.length)

    def __repr__(self) -> str:
        clazz = self.__class__.__name__
        return '<%s: head=%s body_len=%d />' % (clazz, self.head, self.body.length)

    @property
    def head(self) -> Header:
        return self.__head

    @property
    def body(self) -> Data:
        return self.__body

    @classmethod
    def parse(cls, data: Data):  # -> Package
        # get package head
        head = Header.parse(data=data)
        if head is None:
            # raise AssertionError('package head error')
            return None
        # check lengths
        data_len = data.length
        head_len = head.length
        body_len = head.body_length
        if body_len < 0:
            # unlimited
            body_len = data_len - head_len
        pack_len = head_len + body_len
        if data_len < pack_len:
            # raise ValueError('package length error: %s' % data)
            return None
        elif data_len > pack_len:
            data = data.slice(end=pack_len)
        # get body
        if body_len == 0:
            body = Data.ZERO
        else:
            body = data.slice(start=head_len)
        return cls(data=data, head=head, body=body)

    @classmethod
    def new(cls, data_type: DataType, sn: TransactionID = None, pages: int = 1, offset: int = 0,
            body_length: int = -1, body: Data = None):
        # create package with header
        head = Header.new(data_type=data_type, sn=sn, pages=pages, offset=offset, body_length=body_length)
        return cls(head=head, body=body)

    def split(self) -> list:
        """
        Split large message package

        :return: message fragment packages
        """
        head = self.head
        body = self.body
        # check data type
        assert head.data_type == Message, 'cannot split this type: %s' % head.data_type
        # split body
        fragments = []
        count = 1
        start = 0
        end = self.OPTIMAL_BODY_LENGTH
        body_len = body.length
        while end < body_len:
            fragments.append(body.slice(start=start, end=end))
            start = end
            end += self.OPTIMAL_BODY_LENGTH
            count += 1
        if start > 0:
            fragments.append(body.slice(start=start))  # the tail
        else:
            fragments.append(body)
        # create packages with fragments
        data_type = MessageFragment
        sn = head.sn
        packages = []
        if head.body_length < 0:
            # UDP (unlimited)
            for i in range(count):
                body = fragments[i]
                pack = self.new(data_type=data_type, sn=sn, pages=count, offset=i, body_length=-1, body=body)
                packages.append(pack)
        else:
            # TCP (should not happen)
            for i in range(count):
                body = fragments[i]
                pack = self.new(data_type=data_type, sn=sn, pages=count, offset=i, body_length=body.length, body=body)
                packages.append(pack)
        return packages

    @classmethod
    def sort(cls, packages: list) -> list:
        packages.sort(key=lambda pack: pack.head.offset)
        return packages

    @classmethod
    def join(cls, packages: list):  # -> Optional[Package]:
        """
        Join sorted packages' body data together

        :param packages: packages sorted by offset
        :return: original message package
        """
        count = len(packages)
        assert count > 1, 'packages count error: %d' % count
        first = packages[0]
        assert isinstance(first, Package), 'first package error: %s' % first
        trans_id = first.head.sn
        # get fragments count
        pages = first.head.pages
        assert pages == count, 'pages error: %d, %d' % (pages, count)
        # add message fragments part by part
        index = 0
        data = bytearray()
        for item in packages:
            assert isinstance(item, Package), 'package error: %s' % item
            assert item.head.data_type == MessageFragment, 'data type not fragment: %s' % item
            assert item.head.sn == trans_id, 'transaction ID not match: %s' % item
            assert item.head.pages == pages, 'pages error: %s' % item
            assert item.head.offset == index, 'fragment missed: %d, %s' % (index, item)
            data += item.body.get_bytes()
            index += 1
        assert index == pages, 'fragment error: %d/%d' % (index, pages)
        # OK
        body = Data(data=data)
        if first.head.body_length < 0:
            # UDP (unlimited)
            body_len = -1
        else:
            # TCP (should not happen)
            body_len = body.length
        return cls.new(data_type=Message, sn=trans_id, body_length=body_len, body=body)

File: dmtp-java/MTP/src/main/java/chat/dim/mtp/protocol/Package.java
/* license: https://mit-license.org
 *
 *  MTP: Message Transfer Protocol
 *
 *                                Written in 2020 by Moky <albert.moky@gmail.com>
 *
 * ==============================================================================
 * The MIT License (MIT)
 *
 * Copyright (c) 2020 Albert Moky
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 * ==============================================================================
 */
package chat.dim.mtp.protocol;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

import chat.dim.type.ByteArray;
import chat.dim.type.Data;
import chat.dim.type.MutableData;

public class Package extends Data {

    /*    Optimal Length for UDP package body
     *    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     *    MTU      : 576 bytes
     *    IP Head  : 20 bytes
     *    UDP Head : 8 bytes
     *    Header   : 12 bytes (excludes 'pages', 'offset' and 'bodyLen')
     *    Reserved : 24 bytes (includes 'pages', 'offset' and 'bodyLen')
     */
    public static int OPTIMAL_BODY_LENGTH = 512;

    public final Header head;
    public final ByteArray body;

    public Package(ByteArray data, Header head, ByteArray body) {
        super(data);
        this.head = head;
        this.body = body;
    }

    public Package(Header head, ByteArray body) {
        this(head.concat(body), head, body);
    }

    public Package(Header head) {
        this(head, head, Data.ZERO);
    }

    private int getFragmentOffset() {
        return head.offset;
    }

    /**
     *  Split large message package
     *
     * @return message fragment packages
     */
    public List<Package> split() {
        assert head.type.equals(DataType.Message) : "cannot split this type: " + head.type;
        // split body
        List<ByteArray> fragments = new ArrayList<>();
        int count = 1;
        int start = 0, end = OPTIMAL_BODY_LENGTH;
        int length = body.getLength();
        for (; end < length; start = end, end += OPTIMAL_BODY_LENGTH) {
            fragments.add(body.slice(start, end));
            count += 1;
        }
        if (start > 0) {
            fragments.add(body.slice(start)); // the tail
        } else {
            fragments.add(body);
        }
        // create packages with fragments
        List<Package> packages = new ArrayList<>();
        DataType type = DataType.MessageFragment;
        TransactionID sn = head.sn;
        ByteArray data;
        if (head.bodyLength < 0) {
            // UDP (unlimited)
            for (int index = 0; index < count; ++index) {
                data = fragments.get(index);
                packages.add(create(type, sn, count, index, -1, data));
            }
        } else {
            // TCP (should not happen)
            for (int index = 0; index < count; ++index) {
                data = fragments.get(index);
                packages.add(create(type, sn, count, index, data.getLength(), data));
            }
        }
        return packages;
    }

    /**
     *  Join sorted packages' body data together
     *
     * @param packages - packages sorted by offset
     * @return original message package
     */
    public static Package join(List<Package> packages) {
        int count = packages.size();
        assert count > 1 : "packages count error: " + count;
        Package first = packages.get(0);
        DataType type = DataType.MessageFragment;
        TransactionID sn = first.head.sn;
        // get fragments count
        int pages = first.head.pages;
        assert pages == count : "pages error: " + pages + ", " + count;
        // add message fragments part by part
        List<ByteArray> fragments = new ArrayList<>();
        int length = 0;
        int index;
        Package item;
        for (index = 0; index < count; ++index) {
            item = packages.get(index);
            assert type.equals(item.head.type) : "data type should be fragment: " + item;
            assert sn.equals(item.head.sn) : "transaction ID not match: " + item;
            assert pages == item.head.pages : "pages error: " + item;
            assert index == item.head.offset : "fragment missed: " + index;
            fragments.add(item.body);
            length += item.body.getLength();
        }
        assert index == pages : "fragment error: " + index + ", " + pages;
        // join fragments
        MutableData data = new MutableData(length);
        for (index = 0; index < count; ++index) {
            data.append(fragments.get(index));
        }
        type = DataType.Message;
        int bodyLen;
        if (first.head.bodyLength < 0) {
            // UDP (unlimited)
            bodyLen = -1;
        } else {
            // TCP (should not happen)
            bodyLen = data.getLength();
        }
        return create(type, sn, 1, 0, bodyLen, data);
    }

    public static List<Package> sort(List<Package> packages) {
        packages.sort(Comparator.comparingInt(Package::getFragmentOffset));
        return packages;
    }

    public static Package parse(ByteArray data) {
        // get package head
        Header head = Header.parse(data);
        if (head == null) {
            //throw new NullPointerException("package head error: " + Arrays.toString(data));
            return null;
        }
        // check lengths
        int dataLen = data.getLength();
        int headLen = head.getLength();
        int bodyLen = head.bodyLength;
        if (bodyLen < 0) {
            // unlimited
            bodyLen = dataLen - headLen;
        }
        int packLen = headLen + bodyLen;
        if (dataLen < packLen) {
            //throw new ArrayIndexOutOfBoundsException("package length error: " + Arrays.toString(data));
            return null;
        } else if (dataLen > packLen) {
            data = data.slice(0, packLen);
        }
        // get body
        ByteArray body;
        if (bodyLen == 0) {
            body = Data.ZERO;
        } else {
            body = data.slice(headLen);
        }
        return new Package(data, head, body);
    }

    //
    //  Factories
    //

    public static Package create(DataType type, TransactionID sn, int pages, int offset, int bodyLen, ByteArray body) {
        assert body != null : "package body should not be null";
        // create package with header
        Header head = Header.create(type, sn, pages, offset, bodyLen);
        ByteArray data;
        if (body.getLength() > 0) {
            data = head.concat(body);
        } else {
            data = head;
        }
        return new Package(data, head, body);
    }

    //
    //  UDP
    //

    public static Package create(DataType type, TransactionID sn, int pages, int offset, ByteArray body) {
        return create(type, sn, pages, offset, -1, body);
    }

    public static Package create(DataType type, int pages, int offset, ByteArray body) {
        return create(type, TransactionID.generate(), pages, offset, -1, body);
    }

    public static Package create(DataType type, TransactionID sn, ByteArray body) {
        return create(type, sn, 1, 0, -1, body);
    }

    public static Package create(DataType type, ByteArray body) {
        return create(type, TransactionID.generate(), 1, 0, -1, body);
    }

    //
    //  TCP
    //

    public static Package create(DataType type, TransactionID sn, int bodyLen , ByteArray body) {
        assert bodyLen == body.getLength() : "body length error: " + bodyLen + ", " + body.getLength();
        return create(type, sn, 1, 0, bodyLen, body);
    }

    public static Package create(DataType type, int bodyLen, ByteArray body) {
        assert bodyLen == body.getLength() : "body length error: " + bodyLen + ", " + body.getLength();
        return create(type, TransactionID.generate(), 1, 0, bodyLen, body);
    }
}
